include/circt/Dialect/HW/HWAttributesNaming.td

```diff
@@ -17,32 +17,71 @@ include "circt/Dialect/HW/HWDialect.td"
 include "mlir/IR/AttrTypeBase.td"
 
 def InnerRefAttr : AttrDef<HWDialect, "InnerRef"> {
-  let summary = "Refer to a name inside a module";
+  let summary = "Refer to a name inside a list of scopes";
   let description = [{
-    This works like a symbol reference, but to a name inside a module.
+    This works like a symbol reference, but to a name inside a list of `InnerSymbolTable`'s.
   }];
   let mnemonic = "innerNameRef";
-  let parameters = (ins "::mlir::FlatSymbolRefAttr":$moduleRef,
-                        "::mlir::StringAttr":$name);
+  let parameters = (ins ArrayRefParameter<"::mlir::StringAttr">:$path);
 
   let builders = [
-    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$module,
-                                        "::mlir::StringAttr":$name), [{
-      return $_get(
-          module.getContext(), mlir::FlatSymbolRefAttr::get(module), name);
-    }]>
+
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr), [{
+      assert(!pathAttr.empty() && "Empty path");
+      assert(pathAttr.front() && "Null path element");
+      return InnerRefAttr::get(pathAttr.front().getContext(), pathAttr);
+    }]>,
+  
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr,
+                                        "::mlir::StringAttr":$leaf), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath(pathAttr.begin(), pathAttr.end());
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(leaf.getContext(), fullPath);
+    }]>,
+
+    // @mortbopet: Old InnerRef single-module compatability builders.
+    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$leaf), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath;
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+
+    AttrBuilderWithInferredContext<(ins 
+                                    "::mlir::StringAttr":$parent,
+                                    "::mlir::StringAttr":$leaf), [{
+      ::llvm::SmallVector<::mlir::StringAttr> fullPath{parent};
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+    AttrBuilder<(ins "::mlir::StringAttr":$parent,
+                     "::mlir::StringAttr":$leaf), [{
+      return InnerRefAttr::get(parent, leaf);
+    }]>,
   ];
 
   let hasCustomAssemblyFormat = 1;
+  let genVerifyDecl = 1;
 
   let extraClassDeclaration = [{
-    /// Get the InnerRefAttr for an operation and add the sym on it.
-    static InnerRefAttr getFromOperation(mlir::Operation *op,
-                                         mlir::StringAttr symName,
-                                         mlir::StringAttr moduleName);
+    ::mlir::StringAttr getRootReference() const {
+      assert(!getPath().empty() && "Empty path");
```

Comment:
Should the verifier have ensured this?

Reply:
Things which are correct by construction do have a cost in having extra assertions.  Both in debug time and human readability.  The constructor for the attribute should be where the assertion is.